[CA-1058] update rawls-model

andy7i opened the pull request:
\<your comments for this PR go here\>

This PR contains a rawls-model update which led to:
- code changes
- unit test changes
- fixes for sbt assembly merge issues

verified locally using a call to cloneWorkspace with `noWorkspaceOwner` which did not work with the old rawls-model.
https://bvdp-saturn-dev.appspot.com/#workspaces/aj-test-noworkspaceowner-bp/test-localorch-cloneWorkspace--noWorkspaceOwner-true-10-23



Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

andy7i committed:
minor text changes

andy7i committed:
Update rawls-model and fix compile errors due to updated WorkspaceModel which uses Options (see below).

```
case class WorkspaceResponse(accessLevel: Option[WorkspaceAccessLevel],
                             canShare: Option[Boolean],
                             canCompute: Option[Boolean],
                             catalog: Option[Boolean],
                             workspace: WorkspaceDetails,
                             workspaceSubmissionStats: Option[WorkspaceSubmissionStats],
                             bucketOptions: Option[WorkspaceBucketOptions],
                             owners: Option[Set[String]])
```

andy7i commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryServiceSupport.scala:
```diff
@@ -69,11 +69,11 @@ trait LibraryServiceSupport extends DataUseRestrictionSupport with LazyLogging {
-            val newAttrs = replaceDataUseAttributes(ws.attributes, orspAttrs)
-            ws.copy(attributes = newAttrs)
+            val newAttrs = replaceDataUseAttributes(ws.attributes.getOrElse(Map.empty), orspAttrs)
+            ws.copy(attributes = Option(newAttrs))
```
`Option(newAttrs))` or `Some(newAttrs))` ?

MatthewBemis replied:
The way you originally had it is generally preferred because it safely handles null values. `Option(null)` will equate to `None`, but `Some(null)` will throw an exception. I pretty much always use `Option(foo)` now because of this, even if I know that `foo` isn't going to be null

andy7i replied:
good to know -- thanks for the info!

andy7i committed:
readme update

andy7i commented on src/main/scala/org/broadinstitute/dsde/firecloud/utils/PermissionsSupport.scala:
```diff
@@ -49,7 +49,7 @@ trait PermissionsSupport {
       if (!isadmin) {
         rawlsDAO.getWorkspace(workspaceNamespace, workspaceName)(userInfo.asInstanceOf[WithAccessToken]) map { ws =>
-          ws.accessLevel >= neededLevel
+          ws.accessLevel.getOrElse(WorkspaceAccessLevels.NoAccess) >= neededLevel
```
Not 100% sure about this

MatthewBemis replied:
i think it's a fairly safe assumption to make, but a more idiomatic approach may be to use pattern matching:

```
          ws.accessLevel match {
            case Some(accessLevel) => accessLevel >= neededLevel
            case None => false
          }
```

pattern matching may also be useful in some other areas here too if you want to short-circuit any sort of evaluation that would be done by supplying a default in `getOrElse`. sorry i didn't mention this earlier when we were slacking!

andy7i replied:
totally! thanks for the help!

andy7i reviewed (commented):
self-review

andy7i committed:
fix all of the tests

Co-authored-by: s-rubenstein <blrubenstein@wpi.edu>

andy7i committed:
found something similar on github: https://stackoverflow.com/questions/25144484/sbt-assembly-deduplication-found-error

but is that too expansive?

errors for reference:

```
[error] 27 errors were encountered during merge
[trace] Stack trace suppressed: run last *:assembly for the full output.
[error] (*:assembly) deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/ActivationDataFlavor.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/ActivationDataFlavor.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/CommandInfo.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/CommandInfo.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/CommandMap.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/CommandMap.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/CommandObject.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/CommandObject.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataContentHandler.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataContentHandler.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataContentHandlerFactory.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataContentHandlerFactory.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataHandler$1.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataHandler$1.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataHandler.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataHandler.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataHandlerDataSource.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataHandlerDataSource.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataSource.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataSource.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/DataSourceDataContentHandler.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/DataSourceDataContentHandler.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/FileDataSource.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/FileDataSource.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/FileTypeMap.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/FileTypeMap.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/MailcapCommandMap.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/MailcapCommandMap.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/MimeType.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/MimeType.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/MimeTypeParameterList.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/MimeTypeParameterList.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/MimeTypeParseException.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/MimeTypeParseException.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/MimetypesFileTypeMap.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/MimetypesFileTypeMap.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/ObjectDataContentHandler.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/ObjectDataContentHandler.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/SecuritySupport$1.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/SecuritySupport$1.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/SecuritySupport$2.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/SecuritySupport$2.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/SecuritySupport$3.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/SecuritySupport$3.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/SecuritySupport$4.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/SecuritySupport$4.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/SecuritySupport$5.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/SecuritySupport$5.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/SecuritySupport.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/SecuritySupport.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/URLDataSource.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/URLDataSource.class
[error] deduplicate: different file contents found in the following:
[error] /Users/ajang/.ivy2/cache/jakarta.activation/jakarta.activation-api/jars/jakarta.activation-api-1.2.1.jar:javax/activation/UnsupportedDataTypeException.class
[error] /Users/ajang/.ivy2/cache/javax.activation/activation/jars/activation-1.1.jar:javax/activation/UnsupportedDataTypeException.class
[error] Total time: 8 s, completed Oct 23, 2020 6:13:17 PM

```

andy7i commented on project/Merging.scala:
```diff
@@ -5,6 +5,7 @@ object Merging {
     case x if x.endsWith("io.netty.versions.properties") => MergeStrategy.discard
     // we target Java 8, which does not use module-info.class. Some dependencies (Jackson) cause assembly problems on module-info
     case x if x.endsWith("module-info.class") => MergeStrategy.discard
+    case x if x.contains("javax/activation") => MergeStrategy.first
```
how much would it matter if we replaced this section:
```
{
    case x if x.endsWith("io.netty.versions.properties") => MergeStrategy.discard
    // we target Java 8, which does not use module-info.class. Some dependencies (Jackson) cause assembly problems on module-info
    case x if x.endsWith("module-info.class") => MergeStrategy.discard
    case x if x.contains("javax/activation") => MergeStrategy.first
    case x => oldStrategy(x)
  }
```

with:

```
{
 case x => MergeStrategy.first
}
```
https://stackoverflow.com/questions/25144484/sbt-assembly-deduplication-found-error

MatthewBemis replied:
It's pretty rare that these duplicates pop up so personally I prefer dealing with them on a case by case basis rather than handling every case silently. I think it could lead to some tricky debugging if the merge strategy chooses something that the developer isn't expecting

davidangb replied:
+1 to Matt's sentiment!

andy7i committed:
cleanup, early feedback

s-rubenstein reviewed (approved):
We paired on a lot of this, but it looks good to me. I don't have an answer to your question though.

andy7i committed:
update readme on running integration tests locally via intellij

andy7i committed:
PR feedback --

```
Option(null) will equate to None, but Some(null) will throw an exception. I pretty much always use Option(foo) now because of this, even if I know that foo isn't going to be null
```

MatthewBemis reviewed (approved):
:+1:
